pandas/core/base.py

```diff
@@ -735,13 +736,34 @@ def argmin(
                 delegate, skipna=skipna
             )
 
-    def tolist(self):
+    # The overloads are done this way due to a mypy issue. See
+    # https://github.com/python/mypy/issues/3737#issuecomment-465355737
+    @overload
+    def tolist(self) -> list[Any]:
+        ...
+
+    @overload
+    def tolist(self, scalar_type: type[Ttolist]) -> list[Ttolist]:
```

Comment:
Maybe make `scalar_type` keyword-only?

I'm not sure about this change, I understand the intentions but I don't particularly like the mismatch between type-checking and runtime behavior. If people already need to add this additional argument, they could alternative use a cast statement (which is only slightly more verbose).

Reply:
So it becomes about the differences of writing (using a `columns` example):
```python
# What people usually write - but you get no type checking on the `in` operation
cols = [c for c in df.columns if "at" in c]
# versus what would work today where you do get type checking
cols = [c for c in cast(list[str], df.columns.tolist()) if "at" in c]
# versus from this PR where you would get type checking
cols = [c for c in df.columns.tolist(str) if "at" in c]
```

But I see your point - it is only typing 11 more characters.

I did think of adding something to the API, which would be a `get_columns()` method where you passed the type of the `Index`, using something like this:
```python
ScalarT = TypeVar("ScalarT", bound=Scalar)

class ColIndex(pd.Index, Generic[ScalarT]):
    def __iter__(self) -> Iterator[ScalarT]:
        return cast(Iterator[ScalarT], super().__iter__())

# Inside `pd.DataFrame` class:

    def get_columns(coltype: type[ScalarT]) -> ColIndex[ScalarT]:
        return cast(ColIndex[ScalarT], self.columns)
```

Then you could write:
```python
cols = [c for c in df.get_columns(str) if "at" in c]
```
which is a way to get the type of the elements within the `Index`.  

Should I propose that instead?

Reply:
I would be inclined to leave it as-is (`pd.Index.__iter__: Iterator[Any]`) and "wait/hope" for a Generic index in the (far) future.